cannon/multicannon/exec.go

```diff
@@ -0,0 +1,73 @@
+package main
+
+import (
+	"embed"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/versions"
+)
+
+// use the all directive to ensure the .gitkeep file is retained and avoid compiler errors
+
+//go:embed all:embeds
+var vmFS embed.FS
+
+const baseDir = "embeds"
+
+func ExecuteCannon(args []string, ver versions.StateVersion) error {
+	switch ver {
+	case versions.VersionSingleThreaded, versions.VersionMultiThreaded:
+	default:
+		return errors.New("unsupported verrsion")
+	}
+
+	cannonProgramName := vmFilename(ver)
+	cannonProgramBin, err := vmFS.ReadFile(cannonProgramName)
+	if err != nil {
+		return err
+	}
+	cannonProgramPath, err := extractTempFile(filepath.Base(cannonProgramName), cannonProgramBin)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error extracting %s: %v\n", cannonProgramName, err)
+		os.Exit(1)
+	}
+	defer os.Remove(cannonProgramPath)
+
+	if err := os.Chmod(cannonProgramPath, 0755); err != nil {
+		fmt.Fprintf(os.Stderr, "Error setting execute permission for %s: %v\n", cannonProgramName, err)
+		os.Exit(1)
+	}
+
+	execArgs := append([]string{cannonProgramName}, args...)
+
+	// nosemgrep: go.lang.security.audit.dangerous-syscall-exec.dangerous-syscall-exec
+	if err := syscall.Exec(cannonProgramPath, execArgs, os.Environ()); err != nil {
+		fmt.Fprintf(os.Stderr, "Error executing %s: %v\n", cannonProgramName, err)
+		os.Exit(1)
+	}
+
+	panic("unreachable")
+}
+
+func extractTempFile(name string, data []byte) (string, error) {
+	tempDir := os.TempDir()
+	tempFile, err := os.CreateTemp(tempDir, name+"-*")
+	if err != nil {
```

Comment:
One catch here is that a lot of temp directories have the no exec flag set so you can't run an executable from there.  We can test that for our docker image and I suspect it's fine, but something to bear in mind.  We may need to add a flag to control where the binaries are written to. Tempted to defer that until its actually needed though.

Reply:
This works for our docker image on my machine and I think on infra. Though someone could override the default docker tmpfs option with noexec. Will fix if that becomes an issue.
This also made me realize that the temp files are not being cleaned due to the exec. Fixing.